src/sentry/taskworker/scheduler.py

```diff
@@ -0,0 +1,255 @@
+
+    def remaining_delta(self, last_run: datetime | None = None) -> timedelta:
+        """Get a timedelta remaining until the next task should spawn"""
+        if last_run is None:
+            return timedelta(seconds=0)
+        # floor to timestamp as microseconds are not relevant
+        now = int(timezone.now().timestamp())
+        last_run_ts = int(last_run.timestamp())
+
+        seconds_remaining = self._delta.total_seconds() - (now - last_run_ts)
+        return timedelta(seconds=max(seconds_remaining, 0))
+
+    def runtime_after(self, start: datetime) -> datetime:
+        """Get the next time a task should run after start"""
+        return start + self._delta
+
+
+class ScheduleEntry:
+    """Metadata about a task that can be scheduled"""
+
+    def __init__(self, *, task: Task[Any, Any], schedule: timedelta | crontab) -> None:
+        self._task = task
+        scheduler: Schedule
+        if isinstance(schedule, timedelta):
+            scheduler = TimedeltaSchedule(schedule)
+        else:
+            scheduler = CrontabSchedule(schedule)
+        self._schedule = scheduler
+        self._last_run: datetime | None = None
+
+    def __lt__(self, other: ScheduleEntry) -> bool:
+        # Secondary sorting for heapq when remaining time is the same
+        return self.fullname < other.fullname
+
+    @property
+    def fullname(self) -> str:
+        return self._task.fullname
+
+    def set_last_run(self, last_run: datetime | None) -> None:
+        self._last_run = last_run
+
+    def is_due(self) -> bool:
+        return self._schedule.is_due(self._last_run)
+
+    def remaining_delta(self) -> timedelta:
+        return self._schedule.remaining_delta(self._last_run)
+
+    def runtime_after(self, start: datetime) -> datetime:
+        return self._schedule.runtime_after(start)
+
+    def delay_task(self) -> None:
+        self._task.delay()
+
+
+class ScheduleSet:
+    """A collection of ScheduleEntry objects"""
+
+    def __init__(self, registry: TaskRegistry, run_storage: RunStorage) -> None:
+        self._entries: list[ScheduleEntry] = []
+        self._registry = registry
+        self._run_storage = run_storage
+        self._heap: list[tuple[float, ScheduleEntry]] = []
+
+    def add(self, task_config: ScheduleConfig) -> None:
+        """Add a task to the schedule."""
+        try:
+            (namespace, taskname) = task_config["task"].split(":")
+        except ValueError:
+            raise ValueError("Invalid task name. Must be in the format namespace:taskname")
+
+        task = self._registry.get_task(namespace, taskname)
+        entry = ScheduleEntry(task=task, schedule=task_config["schedule"])
+        self._entries.append(entry)
+        self._heap = []
+
+    def tick(self) -> float:
+        """
+        Check if any tasks are due to run at current_time, and spawn them.
+
+        Returns the number of seconds to sleep until the next task is due.
+        """
+        self._build_heap()
```

Comment:
I'm not sure of the use of the heap, if we're just rebuilding it from Redis every time this is called. Why do we need a heap at all, instead of a list of tasks that are due to run?

Reply:
We shouldn't be rebuilding the heap on each tick(). It should only be rebuilt after `add()` clears it. The heap helps in keeping the upcoming tasks sorted by the next spawn time, with the shortest wait time being at the top of the heap.

Reply:
I see, that makes sense. I missed that the build is skipped if the heap already exists.

Reply:
Makes sense. I missed the check for the heap already existing.